core/src/main/scala/org/genivi/sota/core/CampaignLauncher.scala

```diff
@@ -0,0 +1,48 @@
+/**
+  * Copyright: Copyright (C) 2016, ATS Advanced Telematic Systems GmbH
+  * License: MPL-2.0
+  */
+package org.genivi.sota.core
+
+import akka.actor.ActorSystem
+import cats.Foldable
+import cats.implicits._
+import org.genivi.sota.core.data.Campaign
+import org.genivi.sota.core.db.Campaigns
+import org.genivi.sota.common.DeviceRegistry
+import org.genivi.sota.data.{Namespace, PackageId, Uuid}
+import scala.concurrent.{ExecutionContext, Future}
+import slick.driver.MySQLDriver.api.Database
+
+object CampaignLauncher {
+  import UpdateService._
+
+  private def resolve(devices: Seq[Uuid]): DependencyResolver = { pkg =>
+    val elems = devices.map(_ -> Set(pkg.id))
+    Future.successful(Map(elems:_*))
+  }
+
+  private def launchGroup (deviceRegistry: DeviceRegistry,
+                           updateService: UpdateService,
+                           ns: Namespace,
+                           pkgId: PackageId,
+                           groupId: Uuid)
+                          (implicit db: Database, system: ActorSystem, ec: ExecutionContext): Future[Uuid] = for {
+    updateRequest <- updateService.updateRequest(ns, pkgId)
+    devices <- deviceRegistry.fetchGroup(groupId)
+    _ <- updateService.queueUpdate(ns, updateRequest, resolve(devices))
+  } yield Uuid.fromJava(updateRequest.id)
+
+  def launch (deviceRegistry: DeviceRegistry, updateService: UpdateService, id: Campaign.Id)
+             (implicit db: Database, system: ActorSystem, ec: ExecutionContext): Future[List[Uuid]] = for {
+    camp <- db.run(Campaigns.setAsLaunch(id))
+    updateRefs <- camp.groups.map { campGrp =>
+      val groupId = campGrp.group
+      for {
+        uuid <- launchGroup(deviceRegistry, updateService, camp.meta.namespace,
+                            camp.packageId.get, groupId)
+        _ <- db.run(Campaigns.setUpdateUuid(id, groupId, uuid))
+      } yield uuid
+    }.toList.sequence
+  } yield updateRefs
+}
```

Comment:
I tried to make it easier, didn't add any new methods but I did make launchGroup an inner method so that it can share more arguments with launch. The new code doesn't have nested for comprehensions, and I use traverse instead of map(XX).sequence